### ISSUE TITLE
first PublishData always results in an error

### ISSUE THREAD

**Issue by qwertzui11** [livekit/server-sdk-go#49]
Hi! 

Thx for your great work with livekit. It's a really awesome framework and very easy to use! 

However I'm getting an error when sending some data

```go
room.Callback.OnParticipantConnected = func(p *lksdk.RemoteParticipant) {
  if err := room.LocalParticipant.PublishData([]byte("hello"), livekit.DataPacket_RELIABLE, []string{p.SID()}); err != nil {
    log.Printf("could not publish data to participant '%s', err: '%v'", p.SID(), err)
  }
}
```
After a few SECONDS this call results in the error
```
could not publish data to participant 'PA_kCS2E6r2YV3c', err: 'could not connect after timeout'
```
Immediately after that log entry, the following log entry occurs:
```
"level"=0 "msg"="successfully set publisher answer"
```

After that log entry I'm able to successfully send data messages. It seems to me that the first call to `PublishData` blocks some logic.

If I try to send with a go routine like this
```go
	room.Callback.OnParticipantConnected = func(p *lksdk.RemoteParticipant) {
		log.Printf("OnParticipantConnected '%s'", p.Identity())
		go func() {
			time.Sleep(time.Second * 1)
			sendData(p)
			time.Sleep(time.Second * 1)
			sendData(p)
		}()
	}
```
the first send results in the error 
```
could not publish data to participant 'PA_73nWKR3avB7X', err: 'io: read/write on closed pipe'
```
the second send succeeds.

In conclusion: the first `PublishData` always fails. Afterwards it works.

I'm not sure if it matters but the other participant is using the [js-sdk](https://github.com/livekit/client-sdk-js)

Any help would be greatly appreciated  :smile:

**Comment by cnderrauber** [comment]
since publisher is not establish until you publish something(mediatrack or data), so you publish data in OnParticipantConnected will cause publisher pc to connect and block the signal msg process loop, that's why you get timeout.

If you publish data at goroutine, the publisher pc will connected then open the data channel, the publishdata method just wait publish pc connected but data channel may still opening, so your first pub might get closed pipe.

There are two points to improve:
1. we might need make callback in goroutine to avoid user do something block our signal loop @davidzhao 
2. create datachannel after publish pc connected, that might solve your first pub failed problem

**Comment by qwertzui11** [comment]
> 2. create datachannel after publish pc connected, that might solve your first pub failed problem

I'm not sure if I understand this correctly. Can you give me some additional hints how I might `create datachannel after publish pc connected`. What Callback is getting triggered for `pc connected`?

**Comment by davidzhao** [comment]
> 1. we might need make callback in goroutine to avoid user do something block our signal loop @davidzhao
> 2. create datachannel after publish pc connected, that might solve your first pub failed problem

Agreed with both points! Our other client SDKs will block publishData until publish PC is established.

**Comment by cnderrauber** [comment]
@qwertzui11  made pr #50 to fix this, now you can pub data in OnParticipantConnected callback. If you verify all things ok, I'll close this issue

**Comment by qwertzui11** [comment]
Hi! thx for the patch! 

Sadly, no, it does not work. There is no error any more, however the first package is not getting transmitted. The package seems to silently get dropped.

So it does NOT work inside `room.Callback.OnParticipantConnected`. There is no error, however the message gets silently dropped.

The other part is using the javascript sdk. according to `chrome://webrtc-internals/` no message is getting transmitted to the browser. So I assume the go-sdk is at fault.

**Comment by cnderrauber** [comment]
I can receive the first datapacket by use livekit/client-sdk-js's sample app as receiver. send code is like 
```
	room.Callback.OnParticipantConnected = func(rp *lksdk.RemoteParticipant) {
		fmt.Println("conn", rp.Identity())
		err := room.LocalParticipant.PublishData([]byte("123"), livekit.DataPacket_RELIABLE, []string{})
		if err != nil {
			fmt.Println(err)
			panic(err)
		} else {
			fmt.Println("published data")
		}
	}

**Comment by qwertzui11** [comment]
Hi, thx for getting back to me!

I debugged my code and actually, everything seems to work correctly. The Message is getting successfully send on the [`reliableDC`](https://github.com/livekit/server-sdk-go/blob/c873f30833c506a1d8a54066155db1dfed240e04/localparticipant.go#L231)

My code looks like yours, however I'm sending the message to the newly connected participant ONLY.

```go
err := room.LocalParticipant.PublishData([]byte("123"), livekit.DataPacket_RELIABLE, []string{rp.SID()})
```

When I broadcast the message to all participants, it actually works! (`[]string{}`). But when I send it to the newly connected participant the message gets dropped (`[]string{rp.SID()}`).

Due to the lack of errors, I assume it's getting dropped on the [`livekit-server`](https://github.com/livekit/livekit-server)?
I checked the debug error logs of the livekit-server, but there seems to be nothing note worthy.

Thx for your time!
Cheers

**Comment by davidzhao** [comment]
@qwertzui11 which version of livekit-server are you running?

**Comment by qwertzui11** [comment]
@davidzhao 

```bash
root@sfu ~/s/livekit (master)# docker run --rm livekit/livekit-server --version
livekit-server version 0.15.4
```

**Comment by qwertzui11** [comment]
I see, version `0.15.5` is available, I'll re-test within the next 12 hours.

**Comment by cnderrauber** [comment]
I have test with 0.15.4,0.15.5 and master @livekit-server, all of those release can work correctly with send data to special participant
`room.LocalParticipant.PublishData([]byte("123"), livekit.DataPacket_RELIABLE, []string{rp.SID()})`

**Comment by qwertzui11** [comment]
I narrowed the error down to a small project available at https://gitlab.com/qwertzui11/livekit-server-sdk-go-issue-49

the error seems to occur when the `livekit-server` is run without redis

whenever a client leaves the room and then joins again, the first data message is getting dropped. however when everyone leaves the room, and join again, the messages are getting transmitted successfully.

i guess [this issue](https://github.com/livekit/livekit-server/issues/83) succeeds this issue?

Thanks for your time

**Comment by davidzhao** [comment]
@qwertzui11 thank you for the repro

I'm able to reproduce this as well.. It's not clear to me why it's happening though. there seems to be some leftover state here. 

It seems unrelated to #83.

**Comment by qwertzui11** [comment]
Hi, upgraded `livekit-server` to `master` and `server-sdk-go` to `main`. Then re-tested, and success!!!! :heart_eyes: 

Awesome work! Thanks a lot!
